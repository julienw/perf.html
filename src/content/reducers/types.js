// @flow

import type { Summary } from '../../common/summarize-profile';
import type { Action, ExpandedSet, CallTreeFiltersPerThread, DataSource, ProfileSelection } from '../actions/types';
import type { Milliseconds, StartEndRange } from '../../common/types/units';
import type { IndexIntoMarkersTable, IndexIntoFuncTable, Profile, ThreadIndex } from '../../common/types/profile';

export type Reducer<T> = (T, Action) => T;

export type RequestedLib = { pdbName: string, breakpadId: string };
export type SymbolicationStatus = 'DONE' | 'SYMBOLICATING';
export type ThreadViewOptions = {
  selectedFuncStack: IndexIntoFuncTable[],
  expandedFuncStacks: Array<IndexIntoFuncTable[]>,
  selectedMarker: IndexIntoMarkersTable | -1,
};
export type ProfileViewState = {
  viewOptions: {
    threadOrder: number[],
    perThread: ThreadViewOptions[],
    symbolicationStatus: SymbolicationStatus,
    waitingForLibs: Set<RequestedLib>,
    selection: ProfileSelection,
    scrollToSelectionGeneration: number,
    rootRange: StartEndRange,
    zeroAt: Milliseconds,
    tabOrder: number[],
  },
  profile: Profile | null,
};

export type AppState = {
  view: string,
  isURLSetupDone: boolean,
};

export type SummaryViewState = {
  summary: null|Summary,
  expanded: null|ExpandedSet,
};

export type RangeFilterState = {
  start: number,
  end: number,
};

export type URLState = {
  dataSource: DataSource,
  hash: string,
  selectedTab: string,
  rangeFilters: RangeFilterState[],
  selectedThread: ThreadIndex,
  callTreeSearchString: string,
  callTreeFilters: CallTreeFiltersPerThread,
  jsOnly: boolean,
  invertCallstack: boolean,
  hidePlatformDetails: boolean,
}

export type State = {
  app: AppState,
  profileView: ProfileViewState,
  summaryView: SummaryViewState,
  urlState: URLState,
};
