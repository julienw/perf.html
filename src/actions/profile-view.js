/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import {
  selectorsForThread,
  selectedThreadSelectors,
} from '../reducers/profile-view';
import {
  getImplementationFilter,
  getSelectedThreadIndex,
} from '../reducers/url-state';

import type { ProfileSelection, ImplementationFilter } from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoMarkersTable } from '../types/profile';
import type { CallNodePath } from '../types/profile-derived';
import type { Transform } from '../types/transforms';
/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedCallNode(
  threadIndex: ThreadIndex,
  selectedCallNodePath: CallNodePath
): Action {
  return {
    type: 'CHANGE_SELECTED_CALL_NODE',
    selectedCallNodePath,
    threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(
      ({
        type: 'HIDE_THREAD',
        threadIndex,
        threadOrder,
        hiddenThreads,
      }: Action)
    );
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedCallNodes(
  threadIndex: ThreadIndex,
  expandedCallNodePaths: Array<CallNodePath>
): Action {
  return {
    type: 'CHANGE_EXPANDED_CALL_NODES',
    threadIndex,
    expandedCallNodePaths,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker,
    threadIndex,
  };
}

export function changeMarkersSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_MARKER_SEARCH_STRING',
    searchString,
  };
}

export function changeImplementationFilter(
  implementation: ImplementationFilter
): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousImplementation = getImplementationFilter(getState());
    const threadIndex = getSelectedThreadIndex(getState());
    const transformedThread = selectedThreadSelectors.getRangeAndTransformFilteredThread(
      getState()
    );

    dispatch({
      type: 'CHANGE_IMPLEMENTATION_FILTER',
      implementation,
      threadIndex,
      transformedThread,
      previousImplementation,
    });
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(
  hidePlatformDetails: boolean
): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start,
    end,
  };
}

export function addRangeFilterAndUnsetSelection(
  start: number,
  end: number
): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(
      updateProfileSelection({ hasSelection: false, isModifying: false })
    );
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(
  firstRemovedFilterIndex: number
): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(
      updateProfileSelection({ hasSelection: false, isModifying: false })
    );
  };
}

export function addTransformToStack(
  threadIndex: ThreadIndex,
  transform: Transform
): ThunkAction<void> {
  return (dispatch, getState) => {
    const transformedThread = selectorsForThread(
      threadIndex
    ).getRangeAndTransformFilteredThread(getState());

    dispatch({
      type: 'ADD_TRANSFORM_TO_STACK',
      threadIndex,
      transform,
      transformedThread,
    });
  };
}

export function popTransformsFromStack(
  threadIndex: ThreadIndex,
  firstRemovedFilterIndex: number
): Action {
  return {
    type: 'POP_TRANSFORMS_FROM_STACK',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
