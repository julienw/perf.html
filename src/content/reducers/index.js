// @flow
import profileView from './profile-view';
import app from './app';
import summaryView from './summary-view';
import urlState from './url-state';
import flameChart from './flame-chart';
import timelineView from './timeline-view';

const reducer = { app, profileView, summaryView, urlState, flameChart, timelineView };

export default reducer;
