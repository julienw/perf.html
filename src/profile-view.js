/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
/* eslint-disable */
import { createSelector } from 'reselect';

type State = {
  foo: any,
};

const simpleSelector = (state: State): string => 'foo';
const resultSelector = (arg: string): string => 'foo';

const combinedSelector1 = createSelector(
  simpleSelector,
  resultSelector,
);

const combinedSelector2: (state: State) => string = createSelector(
  combinedSelector1,
  resultSelector
);
