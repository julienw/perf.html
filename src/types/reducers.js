/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type {
  Action,
  DataSource,
  ProfileSelection,
  ImplementationFilter,
  TabSlug,
} from './actions';
import type { Milliseconds, StartEndRange } from './units';
import type { IndexIntoMarkersTable, Profile, ThreadIndex } from './profile';
import type { IndexIntoCallNodeTable } from './profile-derived';
import type { Attempt } from '../utils/errors';
import type { GetLabel } from '../profile-logic/labeling-strategies';
import type { GetCategory } from '../profile-logic/color-categories';
import type { TransformStacksPerThread } from './transforms';

export type Reducer<T> = (T, Action) => T;

export type RequestedLib = { debugName: string, breakpadId: string };
export type SymbolicationStatus = 'DONE' | 'SYMBOLICATING';
export type ThreadViewOptions = {
  selectedCallNodeIndex: IndexIntoCallNodeTable | null,
  expandedCallNodeIndexes: Set<IndexIntoCallNodeTable>,
  selectedMarker: IndexIntoMarkersTable | -1,
};
export type ProfileViewState = {
  viewOptions: {
    perThread: ThreadViewOptions[],
    symbolicationStatus: SymbolicationStatus,
    waitingForLibs: Set<RequestedLib>,
    selection: ProfileSelection,
    scrollToSelectionGeneration: number,
    focusCallTreeGeneration: number,
    rootRange: StartEndRange,
    zeroAt: Milliseconds,
    tabOrder: number[],
    rightClickedThread: ThreadIndex,
  },
  profile: Profile,
};

export type AppViewState =
  | {| phase: string |}
  | {
      phase: 'INITIALIZING',
      additionalData: { attempt: Attempt | null, message: string },
    }
  | { phase: 'FATAL_ERROR', error: Error };

export type AppState = {
  view: AppViewState,
  isUrlSetupDone: boolean,
  hasZoomedViaMousewheel: boolean,
};

export type RangeFilterState = {
  start: number,
  end: number,
};

export type UrlState = {
  dataSource: DataSource,
  hash: string,
  profileUrl: string,
  selectedTab: TabSlug,
  rangeFilters: RangeFilterState[],
  selectedThread: ThreadIndex,
  callTreeSearchString: string,
  markersSearchString: string,
  implementation: ImplementationFilter,
  invertCallstack: boolean,
  hidePlatformDetails: boolean,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[],
  transforms: TransformStacksPerThread,
};

export type IconState = Set<string>;

export type StackChartState = {
  categoryColorStrategy: GetCategory,
  labelingStrategy: GetLabel,
};

export type State = {
  app: AppState,
  profileView: ProfileViewState,
  urlState: UrlState,
  stackChart: StackChartState,
  icons: IconState,
};

export type IconWithClassName = {
  icon: string,
  className: string,
};
