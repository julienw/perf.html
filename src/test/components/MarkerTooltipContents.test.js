/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import React from 'react';
import MarkersTooltipContents from '../../components/shared/MarkerTooltipContents';
import renderer from 'react-test-renderer';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithMarkers } from '../fixtures/profiles/make-profile';
import { selectedThreadSelectors } from '../../reducers/profile-view';

describe('MarkerTooltipContents', function() {
  it('renders tooltips for various markers', () => {
    // Enumerate through all of the switch arms of the tooltip for coverage.
    const profile = getProfileWithMarkers([
      [
        'DOMEvent',
        10.5,
        {
          type: 'DOMEvent',
          startTime: 10.5,
          endTime: 11.3,
          eventType: 'commandupdate',
          phase: 2,
        },
      ],
      [
        'UserTiming',
        12.5,
        {
          type: 'UserTiming',
          startTime: 12.5,
          endTime: 12.5,
          name: 'foobar',
          entryType: 'mark',
        },
      ],
      [
        'NotifyDidPaint',
        14.5,
        {
          type: 'tracing',
          startTime: 14.5,
          endTime: 14.5,
          category: 'Paint',
          interval: 'start',
          name: 'NotifyDidPaint',
        },
      ],
      [
        'GCMinor',
        15.5,
        {
          type: 'GCMinor',
          startTime: 15.5,
          endTime: 15.5,
          // nursery is only present in newer profile format.
          nursery: {
            reason: 'CC_WAITING',
            status: 'complete',
            bytes_tenured: 4 * 1024 * 1024,
            bytes_used: 8 * 1024 * 1024,
            new_capacity: 16 * 1024 * 1024,
            lazy_capacity: 12 * 1024 * 1024,
            phase_times: {},
          },
        },
      ],
      [
        'GCMajor',
        16.5,
        {
          type: 'GCMajor',
          startTime: 16.5,
          endTime: 16.5,
          timings: {
            status: 'completed',
            max_pause: 103.65,
            total_time: 157.215,
            reason: 'CC_WAITING',
            zones_collected: 3,
            total_zones: 5,
            total_compartments: 79,
            minor_gcs: 9,
            store_buffer_overflows: 5,
            slices: 13,
            scc_sweep_total: 160,
            scc_sweep_max_pause: 150,
            nonincremental_reason: 'Non incremental reason',
            allocated_bytes: 16 * 1024 * 1024,
            added_chunks: 5,
            removed_chunks: 2,
            major_gc_number: 1,
            minor_gc_number: 2,
            slice_number: 5,
            mmu_20ms: 0,
            mmu_50ms: 0,
            phase_times: {},
          },
        },
      ],
      [
        'GCSlice',
        17.5,
        {
          type: 'GCSlice',
          startTime: 17.5,
          endTime: 17.5,
          timings: {
            reason: 'CC_WAITING',
            slice: 1,
            pause: 5,
            when: 17.5,
            budget: 11,
            initial_state: 'Initial',
            final_state: 'Final',
            major_gc_number: 1,
            page_faults: 0,
            start_timestamp: 17,
            end_timestamp: 17,
            phase_times: {},
          },
        },
      ],
      [
        'Bailout_ShapeGuard after getelem on line 3666 of resource://foo.js -> resource://bar.js:3662',
        10,
        null,
      ],
      ['Invalidate http://mozilla.com/script.js:1234', 10, null],
    ]);
    const { getState } = storeWithProfile(profile);
    const tracingMarkers = selectedThreadSelectors.getTracingMarkers(
      getState()
    );

    tracingMarkers.forEach(marker => {
      expect(
        renderer.create(
          <MarkersTooltipContents
            marker={marker}
            className="propClass"
            threadName="Main Thread"
          />
        )
      ).toMatchSnapshot();
    });
  });
});
