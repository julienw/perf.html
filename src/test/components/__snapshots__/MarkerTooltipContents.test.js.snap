// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerTooltipContents renders tooltips for various markers 1`] = `
Array [
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.000ms
        </div>
        <div
          className="tooltipTitle"
        >
          Bailout
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        Type
        :
      </div>
      ShapeGuard
      <div
        className="tooltipLabel"
      >
        Where
        :
      </div>
      after getelem
      <div
        className="tooltipLabel"
      >
        Script
        :
      </div>
      resource://foo.js -&gt; resource://bar.js
      <div
        className="tooltipLabel"
      >
        Function Line
        :
      </div>
      3662
      <div
        className="tooltipLabel"
      >
        Bailout Line
        :
      </div>
      3666
      <div
        className="tooltipLabel"
      >
        Description:
      </div>
      <div
        style={
          Object {
            "maxWidth": "300px",
          }
        }
      >
        A shape guard based on TI information failed. (We saw an object whose shape does not match that / any of those observed by the baseline IC.)
      </div>
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.000ms
        </div>
        <div
          className="tooltipTitle"
        >
          Invalidate
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        URL
        :
      </div>
      http://mozilla.com/script.js
      <div
        className="tooltipLabel"
      >
        Line
        :
      </div>
      1234
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.80ms
        </div>
        <div
          className="tooltipTitle"
        >
          DOMEvent
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        Type
        :
      </div>
      commandupdate
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.000ms
        </div>
        <div
          className="tooltipTitle"
        >
          UserTiming
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        Name
        :
      </div>
      foobar
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className=""
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          unknown duration
        </div>
        <div
          className="tooltipTitle"
        >
          NotifyDidPaint
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.000ms
        </div>
        <div
          className="tooltipTitle"
        >
          GCMinor
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        Reason
        :
      </div>
      CC_WAITING
      <div
        className="tooltipLabel"
      >
        Bytes tenured
        :
      </div>
      4.0MB / 8.0MB (50%)
      <div
        className="tooltipLabel"
      >
        New nursery size
        :
      </div>
      16MB
      <div
        className="tooltipLabel"
      >
        Lazy-allocated size
        :
      </div>
      12MB
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.000ms
        </div>
        <div
          className="tooltipTitle"
        >
          GCMajor
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        Reason
        :
      </div>
      CC_WAITING
      <div
        className="tooltipLabel"
      >
        Non-incremental reason
        :
      </div>
      Non incremental reason
      <div
        className="tooltipLabel"
      >
        Total slice times
        :
      </div>
      157ms
      <div
        className="tooltipLabel"
      >
        Max Pause
        :
      </div>
      104ms
      <div
        className="tooltipLabel"
      >
        Heap usage
        :
      </div>
      16MB
      <div
        className="tooltipLabel"
      >
        MMU 20ms
        :
      </div>
      0.0%
      <div
        className="tooltipLabel"
      >
        MMU 50ms
        :
      </div>
      0.0%
      <div
        className="tooltipLabel"
      >
        Minor GCs
        :
      </div>
      9
      <div
        className="tooltipLabel"
      >
        Slices
        :
      </div>
      13
      <div
        className="tooltipLabel"
      >
        Zones
        :
      </div>
      3 / 5 (60%)
      <div
        className="tooltipLabel"
      >
        Compartments
        :
      </div>
      79
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.000ms
        </div>
        <div
          className="tooltipTitle"
        >
          GCSlice
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        Reason
        :
      </div>
      CC_WAITING
      <div
        className="tooltipLabel"
      >
        Budget
        :
      </div>
      11
      <div
        className="tooltipLabel"
      >
        States
        :
      </div>
      Initial â€“ Final
      <div
        className="tooltipLabel"
      >
        Page faults
        :
      </div>
      1
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.50ms
        </div>
        <div
          className="tooltipTitle"
        >
          Styles
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetailsBackTrace"
    >
      <h2
        className="tooltipBackTraceTitle"
      >
        First invalidated 
        1.5
        ms before the flush, at:
      </h2>
      <ol
        className="backtrace"
      >
        <li
          className="backtraceStackFrame"
        >
          nsRefreshDriver::AddStyleFlushObserver
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::GeckoRestyleManager::PostRestyleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::GeckoRestyleManager::ContentStateChanged
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::ContentStateChanged
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsDocument::ContentStateChanged
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::dom::Element::RemoveStates
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::UpdateAncestorState
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::SetContentState
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::NotifyMouseOut
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::GenerateMouseEnterExit
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::PreHandleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::HandleEventInternal
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::HandlePositionedEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::HandleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsViewManager::DispatchEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsView::HandleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsChildView::DispatchEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          ChildViewMouseTracker::MouseMoved
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsAppShell::Run
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsAppStartup::Run
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          XREMain::XRE_mainRun
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          XREMain::XRE_main
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          XRE_main
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          _main
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
      </ol>
    </div>
  </div>,
  <div
    className="tooltipMarker propClass"
  >
    <div
      className="tooltipHeader"
    >
      <div
        className="tooltipOneLine"
      >
        <div
          className="tooltipTiming"
        >
          0.50ms
        </div>
        <div
          className="tooltipTitle"
        >
          Styles
        </div>
      </div>
      <div
        className="tooltipDetails"
      >
        <div
          className="tooltipLabel"
        >
          Thread:
        </div>
        Main Thread
      </div>
    </div>
    <div
      className="tooltipDetails"
    >
      <div
        className="tooltipLabel"
      >
        Elements traversed
        :
      </div>
      100
      <div
        className="tooltipLabel"
      >
        Elements styled
        :
      </div>
      50
      <div
        className="tooltipLabel"
      >
        Elements matched
        :
      </div>
      10
      <div
        className="tooltipLabel"
      >
        Styles shared
        :
      </div>
      15
      <div
        className="tooltipLabel"
      >
        Styles reused
        :
      </div>
      20
    </div>
    <div
      className="tooltipDetailsBackTrace"
    >
      <h2
        className="tooltipBackTraceTitle"
      >
        First invalidated 
        0.50
        ms before the flush, at:
      </h2>
      <ol
        className="backtrace"
      >
        <li
          className="backtraceStackFrame"
        >
          nsRefreshDriver::AddStyleFlushObserver
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::GeckoRestyleManager::PostRestyleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::GeckoRestyleManager::ContentStateChanged
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::ContentStateChanged
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsDocument::ContentStateChanged
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::dom::Element::RemoveStates
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::UpdateAncestorState
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::SetContentState
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::NotifyMouseOut
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::GenerateMouseEnterExit
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::EventStateManager::PreHandleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::HandleEventInternal
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::HandlePositionedEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          mozilla::PresShell::HandleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsViewManager::DispatchEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsView::HandleEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsChildView::DispatchEvent
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          ChildViewMouseTracker::MouseMoved
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsAppShell::Run
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          nsAppStartup::Run
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          XREMain::XRE_mainRun
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          XREMain::XRE_main
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          XRE_main
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
        <li
          className="backtraceStackFrame"
        >
          _main
          <em
            className="backtraceStackFrameOrigin"
          >
            
          </em>
        </li>
      </ol>
    </div>
  </div>,
]
`;
